File: backend/app/schemas/social.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, field_validator


# --- Follows ---

class FollowOut(BaseModel):
    id: UUID
    follower_id: UUID
    dog_id: UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class FollowToggle(BaseModel):
    dog_id: UUID


# --- Comments ---

class CommentCreate(BaseModel):
    target_type: str
    target_id: UUID
    body: str

    @field_validator("target_type")
    @classmethod
    def valid_target(cls, v: str) -> str:
        if v not in ("photo", "post", "dog"):
            raise ValueError("target_type must be 'photo', 'post', or 'dog'")
        return v

    @field_validator("body")
    @classmethod
    def body_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Comment body is required")
        if len(v) > 1000:
            raise ValueError("Comment must be 1000 characters or less")
        return v.strip()


class CommentOut(BaseModel):
    id: UUID
    author_id: UUID
    author_name: str | None = None
    target_type: str
    target_id: UUID
    body: str
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Reactions ---

class ReactionToggle(BaseModel):
    target_type: str
    target_id: UUID
    kind: str

    @field_validator("target_type")
    @classmethod
    def valid_target(cls, v: str) -> str:
        if v not in ("photo", "post", "dog"):
            raise ValueError("target_type must be 'photo', 'post', or 'dog'")
        return v

    @field_validator("kind")
    @classmethod
    def valid_kind(cls, v: str) -> str:
        if v not in ("like", "cute", "woof"):
            raise ValueError("kind must be 'like', 'cute', or 'woof'")
        return v


class ReactionCounts(BaseModel):
    like: int = 0
    cute: int = 0
    woof: int = 0
    user_reaction: str | None = None  # What the current user reacted with


# --- User Profile ---

class UserProfileOut(BaseModel):
    id: UUID
    display_name: str
    location_rough: str | None = None
    created_at: datetime
    dog_count: int = 0
    follower_count: int = 0

    model_config = {"from_attributes": True}

File: frontend/src/api/social.ts
import client from './client';

export interface Follow {
  id: string;
  follower_id: string;
  dog_id: string;
  created_at: string;
}

export interface Comment {
  id: string;
  author_id: string;
  author_name: string | null;
  target_type: string;
  target_id: string;
  body: string;
  created_at: string;
}

export interface ReactionCounts {
  like: number;
  cute: number;
  woof: number;
  user_reaction: string | null;
}

export interface UserProfile {
  id: string;
  display_name: string;
  location_rough: string | null;
  created_at: string;
  dog_count: number;
  follower_count: number;
}

// --- Follows ---

export async function followDog(dogId: string): Promise<Follow> {
  const res = await client.post('/social/follows', { dog_id: dogId });
  return res.data;
}

export async function unfollowDog(dogId: string): Promise<void> {
  await client.delete(`/social/follows/${dogId}`);
}

export async function getMyFollows(): Promise<Follow[]> {
  const res = await client.get('/social/follows/mine');
  return res.data;
}

export async function getFollowerCount(dogId: string): Promise<{ count: number; is_following: boolean }> {
  const res = await client.get(`/social/dogs/${dogId}/followers/count`);
  return res.data;
}

// --- Comments ---

export async function createComment(targetType: string, targetId: string, body: string): Promise<Comment> {
  const res = await client.post('/social/comments', { target_type: targetType, target_id: targetId, body });
  return res.data;
}

export async function getComments(targetType: string, targetId: string): Promise<Comment[]> {
  const res = await client.get('/social/comments', { params: { target_type: targetType, target_id: targetId } });
  return res.data;
}

export async function deleteComment(commentId: string): Promise<void> {
  await client.delete(`/social/comments/${commentId}`);
}

// --- Reactions ---

export async function toggleReaction(targetType: string, targetId: string, kind: string): Promise<ReactionCounts> {
  const res = await client.post('/social/reactions', { target_type: targetType, target_id: targetId, kind });
  return res.data;
}

export async function getReactions(targetType: string, targetId: string): Promise<ReactionCounts> {
  const res = await client.get('/social/reactions', { params: { target_type: targetType, target_id: targetId } });
  return res.data;
}

// --- User Profiles ---

export async function getUserProfile(userId: string): Promise<UserProfile> {
  const res = await client.get(`/social/users/${userId}/profile`);
  return res.data;
}
